### PR TITLE
Defer and cache calculation of Engine reporting signature.

### PR DESCRIPTION
**Pull request by abernix**
This moves the Apollo Engine Reporting signature calculation (which the
Apollo Cloud uses to aggregate similar operations) to perform
asynchronously, after the response has been returned to the client, and also
caches the signature for use by additional operations which match the same
`queryHash` (which is an exact SHA-256 hex digest of the operation body
received from the client).

While the signature calculation is relatively quick on small operations, with
sustained load and more complex operations, this recurring calculation can
be more costly.

As a bit of validation to the success of this change, on a very basic
performance benchmark, using a schema with 1000 `String` fields (i.e.  a
`type Query` with `Field_1: String` through `Field_1000: String`) and an
incoming operation which selects from all 1000 of those fields (i.e
`Field_1`), this showed quite an improvement:

```
┌───────────────┬───────────────┬───────────────┐
│               │ Before        │ After         │
├───────────────┼───────────────┼───────────────┤
│ Percentile, % │ Resp. Time, s │ Resp. Time, s │
├───────────────┼───────────────┼───────────────┤
│           0.0 │         0.076 │         0.024 │
│          50.0 │         0.342 │          0.14 │
│          90.0 │         0.388 │         0.161 │
│          95.0 │          0.41 │         0.164 │
│          99.0 │         0.444 │          0.17 │
│          99.9 │         0.486 │         0.177 │
│         100.0 │         0.487 │         0.196 │
└───────────────┴───────────────┴───────────────┘
```

Of course, this is a relatively simple example and still includes actual
GraphQL execution, but the win factor is certainly there.  Other tests with
more dynamic fields and a higher cardinality of unique operations also
showed improvement, though extremely high cardinality of complex operations
(which have more expensive execution characteristics) certainly made the
_win_ factor less pronounced.

It's not surprising that this calculation is a bit expensive since it
requires a number of normalization steps.  Luckily, we can let this all
happen after the request is sent to the client and it's highly cacheable.

By default, we'll use a relatively small cache which should store a large
number of operations, and to avoid introducing configuration surface area
right now without much evidence as to the need, we'll simply log a message
periodically if we're ejecting operations.

### CODE REVIEW

**Comment by abernix** [comment]
This has been published into an `alpha`:

```
Successfully published:
 - apollo-engine-reporting@1.2.0-alpha.0
 - apollo-server-azure-functions@2.6.0-alpha.0
 - apollo-server-cloud-functions@2.6.0-alpha.0
 - apollo-server-cloudflare@2.6.0-alpha.0
 - apollo-server-core@2.6.0-alpha.0
 - apollo-server-express@2.6.0-alpha.0
 - apollo-server-fastify@2.6.0-alpha.0
 - apollo-server-hapi@2.6.0-alpha.0
 - apollo-server-koa@2.6.0-alpha.0
 - apollo-server-lambda@2.6.0-alpha.0
 - apollo-server-micro@2.6.0-alpha.0
 - apollo-server@2.6.0-alpha.0
```

File: packages/apollo-engine-reporting/src/agent.ts [hunk 1]
```diff
@@ -196,7 +212,59 @@ export class EngineReportingAgent<TContext = any> {
...
-  public addTrace(signature: string, operationName: string, trace: Trace) {
+  private async getTraceSignature({
+    queryHash,
+    operationName,
+    documentAST,
+    queryString,
+  }: {
+    queryHash: string;
+    operationName: string;
+    documentAST?: DocumentNode;
```
Comment (martijnwalraven) [inline packages/apollo-engine-reporting/src/agent.ts]: If this is always either a document AST or a string, should this be a single `DocumentNode | string` argument?

File: packages/apollo-engine-reporting/src/agent.ts [hunk 2]
```diff
@@ -196,7 +212,59 @@ export class EngineReportingAgent<TContext = any> {
...
+    if (!documentAST && !queryString) {
+      // This shouldn't happen: one of those options must be passed to runQuery.
+      throw new Error('No queryString or parsedQuery?');
+    }
+
+    // If we didn't have the signature in the cache, we'll resort to
+    // calculating it asynchronously.  The `addTrace` method will
+    // `await` the `signature` if it's a Promise, prior to putting it
+    // on the stack of traces to deliver to the cloud.
+    const cachedSignature = await this.signatureCache.get(queryHash);
```
Comment (martijnwalraven) [inline packages/apollo-engine-reporting/src/agent.ts]: Is the signature calculated per document or per operation? If the latter, we should probably also use `operationName` as part of the key.

File: packages/apollo-engine-reporting/src/agent.ts [hunk 3]
```diff
@@ -351,3 +426,49 @@ export class EngineReportingAgent<TContext = any> {
...
+      // Only show a message warning about the high turnover every 60 seconds.
+      if (
+        !lastSignatureCacheWarn ||
+        new Date().getTime() - lastSignatureCacheWarn.getTime() > 60000
+      ) {
+        // Log the time that we last displayed the message.
+        lastSignatureCacheWarn = new Date();
+        console.warn(
+          [
+            'This server is processing a high number of unique operations.  ',
```
Comment (martijnwalraven) [inline packages/apollo-engine-reporting/src/agent.ts]: Ah, this is a nice touch!

File: packages/apollo-engine-reporting/src/agent.ts (hunk 2 above)
Comment (abernix) [reply packages/apollo-engine-reporting/src/agent.ts]: Great cache-tch!

Comment (abernix) [reply packages/apollo-engine-reporting/src/agent.ts]: Fixed via 5e7084a9cf1056bdf19f63e0a9567701fcc31645.  Thank you!

**Pull request merged**